service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferPolarisEventListener.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.polaris.core.persistence.dao.entity.EntityResult;
+import org.apache.polaris.service.task.TaskExecutor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+/** Event listener that buffers in memory and then dumps to persistence. */
+@ApplicationScoped
+@Identifier("persistence-file-buffer")
+public class FileBufferPolarisEventListener extends PersistencePolarisEventListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileBufferPolarisEventListener.class);
+    MetaStoreManagerFactory metaStoreManagerFactory;
+    PolarisConfigurationStore polarisConfigurationStore;
+    TaskExecutor taskExecutor;
+
+    // Value is a Pair of the first PolarisEvent that is written to the file
+    // (used for checking the first event's timestamp on this shard) and the file handle itself
+    private final HashMap<Integer, BufferShard> buffers = new HashMap<>();
+    private static final int bufferShardCount = 10;
+    private final int timeToFlush;
+
+
+    @Inject
+    public FileBufferPolarisEventListener(
+            MetaStoreManagerFactory metaStoreManagerFactory,
+            PolarisConfigurationStore polarisConfigurationStore,
+            TaskExecutor taskExecutor
+    ) {
+        this.metaStoreManagerFactory = metaStoreManagerFactory;
+        this.polarisConfigurationStore = polarisConfigurationStore;
+        this.taskExecutor = taskExecutor;
+        this.timeToFlush = polarisConfigurationStore.getConfiguration(null, FeatureConfiguration.EVENT_BUFFER_TIME_TO_FLUSH);
+
+        for (int i = 0; i < bufferShardCount; i++) {
+            buffers.put(i, createShard(i));
+        }
+    }
+
+    @Override
+    void addToBuffer(PolarisEvent polarisEvent, CallContext callCtx) {
+        long ingestionTime = System.currentTimeMillis();
+        int shardNum = polarisEvent.hashCode() % bufferShardCount;
+        BufferShard bufferShard = buffers.get(shardNum);
+        if (bufferShard == null) {
+            LOGGER.error("No buffer shard found for shard #{}. Event dropped: {}", shardNum, polarisEvent);
+            return;
+        }
+        if (bufferShard.polarisEvent == null) {
+            bufferShard.polarisEvent = polarisEvent;
+        }
+        try {
+            ObjectOutputStream oos = bufferShard.oos;
+            oos.writeObject(polarisEvent);
+            oos.flush();
+            oos.reset();
+        } catch (IOException e) {
+            LOGGER.error("Unable to write event: {} to buffer shard: {}", polarisEvent, e);
+        }
+
+        if (ingestionTime - bufferShard.polarisEvent.getTimestampMs() > this.timeToFlush) {
+            // Create a new shard (file and writer)
+            buffers.put(shardNum, createShard(shardNum));
+            // Create and fire task to persist the events in the old shard file
```

Comment:
Why go through a task instead of just dumping the buffer file into the DB here?

Reply:
Changed architecture in next revision.